File: algorithms/code/leetcode/lc202_happy_number/lc202_happy_number.py
#!/usr/bin/env python3
# https://leetcode.com/problems/happy-number/

import unittest


class Solution:
    def __re(self, n):
        result = 0
        while n != 0:
            digit = n % 10
            n //= 10
            result += digit * digit
        return result

    def isHappy(self, n: int) -> bool:
        seen = set()
        seen.add(n)
        while n != 1:
            n = self.__re(n)
            if n in seen:
                return False
            seen.add(n)
        return True


class TestCode(unittest.TestCase):
    def test_19(self):
        self.assertTrue(Solution().isHappy(19))

    def test_2(self):
        self.assertFalse(Solution().isHappy(2))

File: algorithms/code/leetcode/lc202_happy_number/lc202_happy_number.rs
// https://leetcode.com/problems/happy-number/
// #easy

pub struct Solution;

impl Solution {
    fn re(mut n: i32) -> i32 {
        let mut result = 0;
        while n != 0 {
            let digit = n % 10;
            n /= 10;
            result += digit * digit;
        }
        result
    }

    pub fn is_happy(mut n: i32) -> bool {
        let mut seen = std::collections::HashSet::new();
        seen.insert(n);
        while n != 1 {
            n = Self::re(n);
            if seen.contains(&n) {
                return false;
            }
            seen.insert(n);
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_19() {
        assert!(Solution::is_happy(19));
    }

    #[test]
    fn test2() {
        assert!(!Solution::is_happy(2));
    }
}
